### PR TITLE
docs: add a new "Glossary" section

### PR DESCRIPTION
**Pull request by cmoussa1**
#### Background

The documentation site contains terms, both Flux-specific and not, that might be unclear to users who are reading the site and/or about Flux for the first time. It was discussed in #185 that the site might benefit from having a page dedicated to explaining in slightly more detail some of the terms, concepts, and acronyms mentioned throughout our site which may be unfamiliar to users who are reading about Flux (or resource managers, HPC concepts, etc.) for the first time.

---

This is a small PR meant to kick off the start of a **Glossary** page for our docs site, a page that explains in more detail these terms and concepts. Mention of this new page is referenced at the top of the **Quick Start** guide for users who might want to look back at this page as they read throughout the guide and other pages on our sites.

A few definitions are also added to this page, but they surely do not make up _all_ of the terms we might want to eventually include. Furthermore, if folks who have feedback/improvements to the terms I've defined in this PR, please let me know. I myself am not familiar with all of the terms suggested in #186 and am pretty sure I did not do all of the definitions their proper justice. 😅 I am simply just trying to start us off!

### CODE REVIEW

File: glossary.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+========
+Glossary
+========
+
+This page contains definitions for Flux-specific terms and concepts mentioned
+throughout the various guides and docs on our sites.
+
+------------------------------------------------------------------------------
+
+**Flux Broker**
```
Comment (vsoch) [inline glossary.rst:12]: Should we have references to each of the terms so they can be linked (specifically) from other pages?

File: glossary.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  A distributed message broker daemon that provides communications services
```
Comment (vsoch) [inline glossary.rst:13]: I would say something in layman's terms like:

> You can imagine the broker as the main controller of the Flux instance - the one running the show, and ensuring that X, Y, Z.

(Metaphors are generally great for docs, when you take a concept and turn it into a mini story so it sticks better).

File: glossary.rst [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  within a Flux instance.
+
+**Flux Instance**
+  A group of flux-broker processes which are launched via any parallel launch
+  utility that supports PMI (Process Manager Interface).
+
+**Hostlist**
```
Comment (vsoch) [inline glossary.rst:20]: Can we show an example here? I would say it's a list of hosts with patterns that are known how to be expanded (and show an example).

File: glossary.rst [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  A format supported by existing HPC tools to create a compact list of hosts by
+  name, consisting of a common prefix, a number, and an optional suffix.
+
+**Jobspec**
```
Comment (vsoch) [inline glossary.rst:24]: Definitely need an example here, even if simple! And then to say what the Jobspec specifically does. E.g., what is creating it? If it's a program, is it being launched somewhere?

File: glossary.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  A file which defines the resources, tasks, and other attributes of a single
+  program.
+
+**Jobtap Plugin**
```
Comment (vsoch) [inline glossary.rst:28]: An example?

File: glossary.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  A built-in or external plugin that can be loaded into the job manager broker
+  module.
+
+**KVS**
```
Comment (vsoch) [inline glossary.rst:32]: Probably should start with "Key Value Store" to say what it means. And add what people use it for (users? developers? both?)

File: glossary.rst [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  A component of a Flux instance where information pertaining to the instance,
+  such as resource information, is stored in a key-value format.
+
+**Limits**
+  Maximum values that are enforced across jobs, queues, and users, such as time
+  and/or resource limits.
+
+**PMI**
```
Comment (vsoch) [inline glossary.rst:40]: What does PMI mean? Why do we use it?

File: quickstart.rst [hunk 8]
```diff
@@ -4,7 +4,9 @@
 Quick Start
 ============
 
-A quick introduction to Flux and flux-core.
+A quick introduction to Flux and flux-core. For reference, the :ref:`Glossary`
```
Comment (vsoch) [inline quickstart.rst:7]: I think this is good for the first shot - what we will eventually do is define a term later and then put a link to it so the person can click and immediately be taken to the exact definition. We can also look into if it might be possible to do some kind of popup so they don't leave the page.

**Review by vsoch — COMMENTED** [review]
Nice!! I left some comments, and let's get others' feedback too.

File: glossary.rst (hunk 2 above)
Comment (cmoussa1) [reply glossary.rst:13]: I like this suggestion a lot, thanks! It might be helpful for me to get an example of what _X,Y,Z_ might be in this metaphor (maybe the flux-core guys can help me out here).

File: glossary.rst (hunk 3 above)
Comment (cmoussa1) [reply glossary.rst:20]: Sure! That sounds like a good suggestion to me. Are you thinking of an example at the end of the definition like the following? "Example: `my_host[1-10]`"

File: glossary.rst (hunk 2 above)
Comment (vsoch) [reply glossary.rst:13]: haha, I put letters because I'm not totally sure. I could guess based on the flux operator, but best to ask. If it's helpful here is the template I use:

```yaml
# Flux needs to know the path to the IMP executable
[exec]
imp = "/usr/libexec/flux/flux-imp"

[access]
allow-guest-user = true
allow-root-owner = true

# Point to resource definition generated with flux-R(1).
[resource]
path = "/etc/flux/system/R"

[bootstrap]
curve_cert = "/etc/curve/curve.cert"
default_port = 8050
default_bind = "tcp://eth0:%%p"
default_connect = "tcp://%%h.%s:%%p"
hosts = [
	{ host="%s-%s"},
]
```

File: glossary.rst (hunk 3 above)
Comment (vsoch) [reply glossary.rst:20]: Yes exactly! Say something like `my_host[1-10]` expands to `my_host1` `my_host2` ... `my_host10`

File: glossary.rst (hunk 5 above)
Comment (cmoussa1) [reply glossary.rst:28]: Hmmm, I wonder if there's an example small enough to paste on this page. I guess I could paste a link to the [flux-accounting jobtap plugin](https://github.com/cmoussa1/flux-accounting/blob/master/src/plugins/mf_priority.cpp) or the examples in the [flux-core repo](https://github.com/flux-framework/flux-core/tree/master/t/job-manager/plugins) (thinking about it out loud, linking the latter might make more sense as those are written in C, like the rest of flux-core). What do you think?

File: glossary.rst (hunk 6 above)
Comment (cmoussa1) [reply glossary.rst:32]: Oh, yes, can't believe I forgot to spell out the acronym (when that's what we talked about why the Glossary page was necessary 🤦).

> And add what people use it for (users? developers? both?)

Good point. Will also ping this to the flux-core guys to get more explanation.

File: glossary.rst (hunk 7 above)
Comment (cmoussa1) [reply glossary.rst:40]: I should spell this out too (PMI stands for (I believe) **Process Manager Interface**). This definition is also something I think I could use some help on from the flux-core guys ^

File: glossary.rst (hunk 5 above)
Comment (vsoch) [reply glossary.rst:28]: I think that would work for now - in the future for developer docs we will want documentation that unwraps what it means to be a plugin, how to write one, etc.

File: glossary.rst (hunk 7 above)
Comment (garlick) [reply glossary.rst:40]: Maybe add:
> PMI traditionally provides a key-value store with all-to-all style synchronization to support the bootstrap parallel programs such as MPI.   During wire up, each parallel task can bind to a network endpoint, put its network address into the PMI KVS, synchronize, and then get the addresses of its network peers.  All without requiring the process manager (e.g. Flux) to know anything about the network interconnect hardware.  Brilliant!
>
> Although there are several variations of PMI, Flux implements the original one from Argonne National Lab which is  documented in [RFC 13](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_13.html).   In addition to providing PMI service to the programs it launches, Flux also _uses_ PMI for its own wire-up when it is launched by Flux or a foreign launcher.
>
> PMI was an unfortunate three letter acronym choice, given that MPI was already a thing.  This is known.

**Comment by chu11** [comment]
I think this is an awesome start.  Once we get this merged, we can all start adding a few more terms here and there.  I'm already gearing for some of my additions!

File: glossary.rst (hunk 4 above)
Comment (garlick) [reply glossary.rst:24]: Maybe some of the terms and definitions listed here could be just slotted in as a first pass?

https://flux-framework.readthedocs.io/en/latest/guides/admin-guide.html#software-components

File: glossary.rst (hunk 3 above)
Comment (grondo) [reply glossary.rst:20]: One or more of the [test vectors](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_29.html#test-vectors) from the RFC might work?

File: glossary.rst (hunk 5 above)
Comment (grondo) [reply glossary.rst:28]: there are some simple jobtap plugins used for testing in the flux-core testsuite. However, I'm not convinced copy and paste of some C code would elucidate much here since even a simple plugin might be quite verbose. Would it be enough to paste a link to the `flux-jobtap-plugins(7)` man page?

**Comment by cmoussa1** [comment]
OK, I just force-pushed some changes that (hopefully) should address most of the great feedback and suggestions I received above. These changes include the following:

* references are now added to every entry on the Glossary page so that they can be linked to throughout the site
* some of the definitions have been reworked to hopefully be more explanatory (thanks for the suggestions on definitions above and in Slack DMs 😄)
* the terms in the Glossary that are acronyms have their abbreviations spelled out when they are mentioned for the first time
* hyperlinks are added to the first mention of some of the Flux terms in the Quickstart page that redirect the user to the definition of the term on the Glossary page

File: glossary.rst [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  IDs used to represent Flux jobs, defined by
+  `RFC 19 <https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_19.html>`_.
+
+.. _hostlist:
+
+**Hostlist**
+  A format supported by existing HPC tools to create a compact list of hosts by
+  name, consisting of a common prefix, a number, and an optional suffix. For
+  example, the hostlist ``"foo[1-5]"`` can be expanded to
+  ``"foo1,foo2,foo3,foo4,foo5"``.
```
Comment (chu11) [inline glossary.rst:46]: instead of "existing", "many"?

Comment (chu11) [reply glossary.rst:46]: Perhaps say "the hosts foo1,foo2, and foo3 could be represented by the hostlist foo[1-3]."?

Comment (vsoch) [reply glossary.rst:46]: Yeah I like that! And maybe go nuts and pick a fun word like fry, cow, cat, bun, fun.... there are lots of good ones! https://www.playosmo.com/kids-learning/3-letter-words-for-kids/ :laughing: 

:potato: :cow: :cat: :bread: :tada: 

Extra points for a matching emoji!  A nicely placed emoji, maybe one per page of docs, makes them more fun.

Comment (cmoussa1) [reply glossary.rst]: Cool! Thanks to you both for the suggestions. I just force-pushed a change that re-worded the example at the end of the definition for hostlist. I chose `cow` for our pretend cluster, lol. 🐮

File: glossary.rst [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  PMI for its own wire-up when it is launched by Flux or a foreign launcher.
+
+  PMI was an unfortunate three letter acronym choice, given that MPI was already a
+  thing. This is known - no whining.
+
+.. _policy:
+
+**Policy**
+  Attributes of a Fux queue that define behavior of the queue, such as time or
+  resource limits.
```
Comment (chu11) [inline glossary.rst:113]: Fux -> Flux

**Review by chu11 — APPROVED** [review]
LGTM, just noticed the one typo below.

File: glossary.rst (hunk 10 above)
Comment (cmoussa1) [reply glossary.rst]: Thanks, just force-pushed a change to fix this typo.

**Comment by cmoussa1** [comment]
Thanks for approving @chu11! @vsoch, did you want to give this a PR another look?

**Review by vsoch — APPROVED** [review]

**Comment by cmoussa1** [comment]
Woo! Thanks. Setting MWP 🎉
